### ISSUE TITLE
Connection lost, unable to rollback after Oracle ORA-00060

### ISSUE THREAD

**Issue by dgrandemange** [brettwooldridge/HikariCP#992]
#### Environment
```
HikariCP version: 2.4.6
JDK version     : 1.8.0_141-b15
Database        : Oracle
Driver version  : ojdbc7-12.1.0.2
Persistence     : JPA2.1, Hibernate 4.3.11
```
-----------------------------------------------------------------------------------------

Hi,

We are using Hikari pool within a Spring Batch application.
Spring Batch job step execution does its work is one transaction. 
So, transaction is committed once the step is successfully done. 
In case an error occurs somewhere in the step processing, exception is raised, which is handled by Spring to rollback the transaction.

I have a case where an error has occurred during step processing, but rollback couldn't be done.
Actually, rollback has been triggered, but has failed because the JDBC connection wasn't available anymore at the time of this rollback attempt.

Her the sequence of events (logs extract) :

    2017-10-03 18:25:28,085  480015 WARN  c.z.hikari.pool.ProxyConnection- springHikariCP - Connection oracle.jdbc.driver.T4CConnection@befb3 marked as broken because of SQLSTATE(61000), ErrorCode(60)
    java.sql.SQLException: ORA-00060: deadlock detected while waiting for resource ...

    2017-10-03 18:25:28,094  480024 ERROR c.o.s.s.c.s.l.ListenerTraitementProcess- ERREUR org.springframework.dao.CannotAcquireLockException: could not execute statement; SQL [n/a]; nested exception is org.hibernate.exception.LockAcquisitionException: could not execute statement

    2017-10-03 18:25:28,095  480025 ERROR o.s.b.core.step.tasklet.TaskletStep- Rolling back with transaction in unknown state
    
    2017-10-03 18:25:28,096  480026 ERROR o.s.t.support.TransactionTemplate- Application exception overridden by rollback exception 
    
    2017-10-03 18:25:28,101  480031 ERROR c.o.s.s.c.s.l.ListenerTraitementStepExecution- [Could not roll back JPA transaction; nested exception is javax.persistence.PersistenceException: unexpected error when rollbacking]
    Caused by: javax.persistence.PersistenceException: unexpected error when rollbacking
    ...
    Caused by: org.hibernate.TransactionException: rollback failed
    ...
    Caused by: org.hibernate.TransactionException: unable to rollback against JDBC connection
    ...
    Caused by: java.sql.SQLException: Connection is closed
    ...

Considering the source cause of this events sequence :
`java.sql.SQLException: ORA-00060: deadlock detected while waiting for resource`, 
could it be that Hikari reacted to this error and closed the JDBC connection on its own ?
Can you see any reason of such a connection loss ?

Thank you for your time

Damien

**Comment by dgrandemange** [comment]
Ok, here is an interesting log lin i fidn't mention in the first place : 

`2017-10-05 14:44:04,088 90027 WARN  c.z.hikari.pool.ProxyConnection - springHikariCP - Connection oracle.jdbc.driver.T4CConnection@19bf812 marked as broken because of SQLSTATE(61000), ErrorCode(60) java.sql.SQLException: ORA-00060: deadlock detected while waiting for resource`

Given `SQLSTATE(61000)`, and after looking at method of [ProxyConnection.checkException(final SQLException sqle)](https://github.com/brettwooldridge/HikariCP/blob/HikariCP-2.4.6/src/main/java/com/zaxxer/hikari/pool/ProxyConnection.java#L149), it seems the connection is evicted because  SQLSTATE 61000 is declared in the [SQL_ERRORS set](https://github.com/brettwooldridge/HikariCP/blob/HikariCP-2.4.6/src/main/java/com/zaxxer/hikari/pool/ProxyConnection.java#L77). And as far as i understand any conneciton in a state within this set is evicted. Is it right ?

In short term, i cannot avoid those deadlock issues (ORA-00060), do yuo by chance have any clue so that Hikari does not evict the connection in such case ?

**Comment by dgrandemange** [comment]
Ok, since [HikariCP-2.4.7](https://github.com/brettwooldridge/HikariCP/blob/HikariCP-2.4.7/CHANGES), a specific handling has been implemented : 

> Removed Oracle SQL state 61000, added specific error code (2399) to evict connections
when it is encountered.
